src/Sarif.ValidationTests/DefaultObjectPopulatingVisitor.cs

```diff
@@ -0,0 +1,293 @@
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+using Microsoft.Json.Schema;
+
+namespace Microsoft.CodeAnalysis.Sarif
+{
+    /// <summary>
+    /// This visitor will exhaustively populate a SARIF log with default values, starting at 
+    /// whatever level in the tree is visited. This has the effect of comprehehsively 
+    /// 'hydrating' all possible SARIF constructs as defined by the C# OM.
+    /// </summary>
+    public class DefaultObjectPopulatingVisitor : SarifRewritingVisitor
+    {
+        public delegate object PrimitiveValueBuilder(bool isRequired);
+
+        // The default builders create a placeholder value for all required property primitives
+        // and emits default (null, 0, false, etc.) values for the remainder. As a result, a
+        // default hydration of this class should contain no non-default values for 
+        // JSON primitive-valued properties that aren't required
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateRequiredPrimitives()
+        {
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return isRequired; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 0; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 0; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "Required string value." : null; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(); } },
+                { typeof(Uri),     (isRequired) => { return isRequired ? new Uri("https://required.uri.contoso.com") : null; } }
+            };
+        }
+
+        // These builders exhaustively populate the generated SARIF file, including non-required JSON primitives
+        public static IDictionary<Type, PrimitiveValueBuilder> PopulateAllPrimitives()
+        {
+            // The values in this output are designed to allow conformance to the schema. In our schema,.
+            // for example, some integer values must be > 0. And so, out optional integer value is 
+            // written as 1.
+            return new Dictionary<Type, PrimitiveValueBuilder>()
+            {
+                { typeof(bool),    (isRequired) => { return true; } },
+                { typeof(int),     (isRequired) => { return isRequired ? int.MaxValue : 1; } },
+                { typeof(double),  (isRequired) => { return isRequired ? double.MaxValue : 1; } },
+                { typeof(string),  (isRequired) => { return isRequired ? "[Required string value]" : "[Optional string value]"; } },
+                { typeof(DateTime),(isRequired) => { return isRequired ? DateTime.UtcNow : new DateTime(1776, 7, 4).ToUniversalTime(); } },
```

Comment:
>DateTime.UtcNow [](start = 73, length = 15)

This code will be used in unit tests, and`DateTime.UtcNow` isn't determinstic. Just use a fixed date, _e.g_ `new DateTime(2018, 10, 29).ToUniversalTime()`.

Reply:
This is not fixed in Iteration 3 (the latest iteration I have). Do you have more changes to push?

---
In reply to: [229106315](https://github.com/Microsoft/sarif-sdk/pull/1121#discussion_r229106315) [](ancestors = 229106315)